Bug: Missing return path in function only caught at runtime, not compile-time

## Description

Functions that declare a return type but don't return a value on all code paths are only caught at runtime when that path is executed, rather than at compile-time during type checking.

## Steps to Reproduce

```ez
import @std
using std

do maybe_return(x int) -> int {
    if x > 0 {
        return x
    }
    // No return here - this path doesn't return anything!
}

do main() {
    temp result1 int = maybe_return(5)   // Works fine
    println("result1 = ${result1}")
    
    temp result2 int = maybe_return(-1)  // Runtime error here
    println("result2 = ${result2}")
}
```

## Expected Behavior

The typechecker should detect that `maybe_return` doesn't return a value on all code paths and report a compile-time error like:

```
error: function 'maybe_return' does not return a value on all code paths
```

## Actual Behavior

The code compiles successfully. When `maybe_return(-1)` is called (hitting the non-returning path), a runtime error occurs:

```
error[E5012]: return type mismatch: expected int, got nil
```

## Why This Matters

- **Delayed feedback**: Users don't discover the bug until that specific code path is executed
- **Testing burden**: Requires comprehensive test coverage to catch these issues
- **Production risk**: Bug could slip into production if that path isn't tested

## Suggested Fix

Add control flow analysis to the typechecker to verify all code paths in functions with return types end with a `return` statement.

--- a/ERRORS.md
+++ b/ERRORS.md
@@ -130,6 +130,7 @@ Type system errors
 | E3032 | enum-type-mismatch | cannot compare values from different enum types |
 | E3033 | duplicate-enum-value | enum contains duplicate values |
 | E3034 | any-type-not-allowed | 'any' type is reserved for internal use |
+| E3035 | not-all-paths-return | not all code paths return a value |
 
 ## Reference Errors (E4xxx)
 

--- a/integration-tests/fail/errors/E3035_not_all_paths_return.ez
+++ b/integration-tests/fail/errors/E3035_not_all_paths_return.ez
@@ -1,0 +1,20 @@
+/*
+ * Error Test: E3035 - not-all-paths-return
+ *
+ * Functions with return types must return a value on ALL code paths.
+ * This function only returns in the if branch, not the fallthrough.
+ */
+import @std
+using std
+
+do maybe_return(x int) -> int {
+    if x > 0 {
+        return x
+    }
+    // Missing return when x <= 0
+}
+
+do main() {
+    temp result int = maybe_return(5)
+    println(result)
+}

--- a/pkg/errors/codes.go
+++ b/pkg/errors/codes.go
@@ -132,6 +132,7 @@ var (
 	E3032 = ErrorCode{"E3032", "enum-type-mismatch", "cannot compare values from different enum types"}
 	E3033 = ErrorCode{"E3033", "duplicate-enum-value", "enum contains duplicate values"}
 	E3034 = ErrorCode{"E3034", "any-type-not-allowed", "'any' type is reserved for internal use"}
+	E3035 = ErrorCode{"E3035", "not-all-paths-return", "not all code paths return a value"}
 )
 
 // =============================================================================

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -916,12 +916,21 @@ func (tc *TypeChecker) checkFunctionBody(node *ast.FunctionDeclaration) {
 		tc.defineVariableWithMutability(param.Name.Value, param.TypeName, param.Mutable)
 	}
 
-	// Check if function body contains at least one return statement (for functions with return types)
+	// Check if function body returns on all code paths (for functions with return types)
 	if len(node.ReturnTypes) > 0 {
 		if !tc.hasReturnStatement(node.Body) {
+			// No return statement at all
 			tc.addError(
 				errors.E3024,
 				fmt.Sprintf("Function '%s' declares return type(s) but has no return statement", node.Name.Value),
+				node.Name.Token.Line,
+				node.Name.Token.Column,
+			)
+		} else if !tc.allPathsReturn(node.Body) {
+			// Has return statements, but not on all code paths
+			tc.addError(
+				errors.E3035,
+				fmt.Sprintf("Function '%s' does not return a value on all code paths", node.Name.Value),
 				node.Name.Token.Line,
 				node.Name.Token.Column,
 			)
@@ -1026,6 +1035,62 @@ func (tc *TypeChecker) hasReturnInIfStatement(ifStmt *ast.IfStatement) bool {
 			// Check the otherwise block
 			return tc.hasReturnStatement(altBlock)
 		}
+	}
+
+	return false
+}
+
+// allPathsReturn checks if ALL code paths in a block return a value.
+// This is stricter than hasReturnStatement which only checks if ANY path returns.
+func (tc *TypeChecker) allPathsReturn(block *ast.BlockStatement) bool {
+	if block == nil || len(block.Statements) == 0 {
+		return false
+	}
+
+	for _, stmt := range block.Statements {
+		switch s := stmt.(type) {
+		case *ast.ReturnStatement:
+			// Found a return at this level - this path returns
+			return true
+
+		case *ast.IfStatement:
+			// For an if statement to guarantee a return on all paths:
+			// 1. It must have an otherwise (else) clause
+			// 2. Both the if branch AND the otherwise branch must all-paths-return
+			if tc.ifAllPathsReturn(s) {
+				return true
+			}
+			// If the if doesn't cover all paths, continue checking subsequent statements
+		}
+		// Loops (for, foreach, while) can't guarantee they execute,
+		// so we can't count returns inside them as covering all paths.
+		// Continue to next statement.
+	}
+
+	// Reached end of block without finding a guaranteed return
+	return false
+}
+
+// ifAllPathsReturn checks if an if/or/otherwise chain returns on ALL paths
+func (tc *TypeChecker) ifAllPathsReturn(ifStmt *ast.IfStatement) bool {
+	// The consequence (if block) must return on all its paths
+	if !tc.allPathsReturn(ifStmt.Consequence) {
+		return false
+	}
+
+	// Must have an alternative (or/otherwise)
+	if ifStmt.Alternative == nil {
+		return false
+	}
+
+	// Check the alternative
+	switch alt := ifStmt.Alternative.(type) {
+	case *ast.IfStatement:
+		// It's an "or" (else if) - recursively check
+		return tc.ifAllPathsReturn(alt)
+	case *ast.BlockStatement:
+		// It's an "otherwise" (else) block
+		return tc.allPathsReturn(alt)
 	}
 
 	return false

--- a/pkg/typechecker/typechecker_test.go
+++ b/pkg/typechecker/typechecker_test.go
@@ -687,6 +687,115 @@ do main() {}
 }
 
 // ============================================================================
+// All Paths Return Tests (E3035)
+// ============================================================================
+
+func TestNotAllPathsReturn_IfWithoutOtherwise(t *testing.T) {
+	input := `
+do maybe_return(x int) -> int {
+	if x > 0 {
+		return x
+	}
+}
+
+do main() {}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E3035)
+}
+
+func TestNotAllPathsReturn_IfOrWithoutOtherwise(t *testing.T) {
+	input := `
+do classify(x int) -> string {
+	if x > 0 {
+		return "positive"
+	} or x < 0 {
+		return "negative"
+	}
+}
+
+do main() {}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E3035)
+}
+
+func TestNotAllPathsReturn_ReturnOnlyInLoop(t *testing.T) {
+	input := `
+do find_first(arr [int]) -> int {
+	for_each i in arr {
+		return i
+	}
+}
+
+do main() {}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E3035)
+}
+
+func TestAllPathsReturn_IfOtherwise(t *testing.T) {
+	input := `
+do get_sign(x int) -> string {
+	if x > 0 {
+		return "positive"
+	} otherwise {
+		return "non-positive"
+	}
+}
+
+do main() {}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestAllPathsReturn_IfOrOtherwise(t *testing.T) {
+	input := `
+do classify(x int) -> string {
+	if x > 0 {
+		return "positive"
+	} or x < 0 {
+		return "negative"
+	} otherwise {
+		return "zero"
+	}
+}
+
+do main() {}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestAllPathsReturn_DirectReturn(t *testing.T) {
+	input := `
+do double(x int) -> int {
+	return x * 2
+}
+
+do main() {}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestAllPathsReturn_ReturnAfterIf(t *testing.T) {
+	input := `
+do abs_value(x int) -> int {
+	if x < 0 {
+		return -x
+	}
+	return x
+}
+
+do main() {}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+// ============================================================================
 // Operator Type Checking Tests
 // ============================================================================
 

